query-service-impl/src/main/java/org/hypertrace/core/query/service/promql/QueryRequestToPromqlConverter.java

```diff
@@ -0,0 +1,224 @@
+import org.hypertrace.core.query.service.api.LiteralConstant;
+import org.hypertrace.core.query.service.api.QueryRequest;
+import org.hypertrace.core.query.service.api.Value;
+import org.hypertrace.core.query.service.pinot.ViewDefinition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class QueryRequestToPromqlConverter {
+
+  private static final Logger LOG = LoggerFactory.getLogger(QueryRequestToPromqlConverter.class);
+
+  private final ViewDefinition viewDefinition;
+  private final Joiner joiner = Joiner.on(", ").skipNulls();
+
+  QueryRequestToPromqlConverter(
+      ViewDefinition viewDefinition) {
+    this.viewDefinition = viewDefinition;
+  }
+
+  /**
+   * 1. selection should be equal to group by list
+   * 2. number of function selection is equal to number of query fired
+   * 3. count(*) type of query can't be served
+   */
+  PromqlQuery toPromql(
+      ExecutionContext executionContext,
+      QueryRequest request,
+      LinkedHashSet<Expression> allSelections) {
+    String groupByList = getGroupByList(request);
+
+    StringBuilder filter = new StringBuilder();
+    convertFilter2String(request.getFilter(), filter);
+
+
+    QueryTimeRange queryTimeRange = executionContext.getQueryTimeRange().orElseThrow(
+        () -> new RuntimeException("Time Range missing in query"));
+
+    List<String> queries = allSelections.stream().filter(expression -> expression.getValueCase().equals(ValueCase.FUNCTION))
+        .map(functionExpression -> {
+          String functionName = getFunctionName(functionExpression);
+          String metricName = getMetricNameFromFunction(functionExpression);
+          return buildQuery(metricName, functionName, groupByList, filter.toString(), queryTimeRange.getDuration().toMillis());
+        }).collect(Collectors.toUnmodifiableList());
+
+    return new PromqlQuery(
+        queries, executionContext.getTimeSeriesPeriod().isEmpty(),
+        queryTimeRange.getStartTime(), queryTimeRange.getEndTime(),
+        getStep(executionContext));
+  }
+
+  private String getGroupByList(QueryRequest queryRequest) {
+    List<String> groupByList = Lists.newArrayList();
+    if (queryRequest.getGroupByCount() > 0) {
+      for (Expression expression : queryRequest.getGroupByList()) {
+        groupByList.add(convertColumnIdentifierExpression2String(expression));
+      }
+    }
+    return String.join(",", groupByList);
+  }
+
+  private long getStep(ExecutionContext executionContext) {
+    return executionContext
+        .getTimeSeriesPeriod().map(Duration::toMillis)
+        .orElse(-1L);
+  }
+
+  private String buildQuery(
+      String metricName, String function,
+      String groupByList, String filter, long durationMillis) {
+    String template = "%s by (%s) (%s(%s{%s}[%sms]))"; // sum by (a1, a2) (sum_over_time(num_calls{a4="..", a5=".."}[xms]))
+
+    return String.format(template, function, groupByList, function + "_over_time", metricName, filter, durationMillis);
```

Comment:
The Range vector selection function (https://prometheus.io/docs/prometheus/latest/querying/basics/#range-vector-selectors) will be based on the type of metrics - `Gauge, Counter, etc`.

So, we should take that in viewDefintion about type of metric too. However, currently, all our generated metrics will be the type of Guage.

Reply:
Added in PrometheusViewDefinition